OpenEdXMobile/src/main/java/org/edx/mobile/view/MyCoursesListFragment.kt

```diff
@@ -192,8 +234,54 @@ class MyCoursesListFragment : OfflineSupportBaseFragment(), RefreshListener {
                     iapAnalytics.trackIAPEvent(Analytics.Events.IAP_UNLOCK_UPGRADED_CONTENT_REFRESH_TIME)
                     SnackbarErrorNotification(binding.root).showError(R.string.purchase_success_message)
                     iapViewModel.resetPurchase(false)
+
+                    if (unfulfilledPurchases.isNotEmpty()) {
+                        purchaseCourse()
+                    }
                 }
             })
+
+        iapViewModel.errorMessage.observe(
+            viewLifecycleOwner,
+            NonNullObserver { errorMsg ->
+                if (iapViewModel.upgradeMode.isNormalMode()) return@NonNullObserver
+
+                if (errorMsg.throwable is InAppPurchasesException) {
+                    when (errorMsg.throwable.httpErrorCode) {
+                        HttpStatus.UNAUTHORIZED -> {
+                            environment.router?.forceLogout(
+                                requireContext(),
+                                environment.analyticsRegistry,
+                                environment.notificationDelegate
+                            )
+                            return@NonNullObserver
+                        }
+                        else -> iapUtils.showPostUpgradeErrorDialog(
```

Comment:
why we have `iapUtils.showPostUpgradeErrorDialog` twice in this observer?
can we make it more efficient?

Reply:
It would make the Error Alerts a bit more complex.